File: src/1352-Product-of-the-Last-K-Numbers/1352.go
type ProductOfNumbers struct {
    pre []int
}


func Constructor() ProductOfNumbers {
    this := new(ProductOfNumbers)
    this.pre = []int{1}
    return *this
}


func (this *ProductOfNumbers) Add(num int)  {
    if num == 0 {
        this.pre = []int{1}
    } else {
        this.pre = append(this.pre, this.pre[len(this.pre) - 1] * num)
    }
}


func (this *ProductOfNumbers) GetProduct(k int) int {
    if k >= len(this.pre) {
        return 0
    }
    return this.pre[len(this.pre) - 1] / this.pre[len(this.pre) - k - 1]
}

File: src/1352-Product-of-the-Last-K-Numbers/1352.java
class ProductOfNumbers {

    public ProductOfNumbers() {
        pre.add(1);
    }
    
    public void add(int num) {
        if (num == 0) {
            pre.clear();
            pre.add(1);
        } else {
            pre.add(pre.get(pre.size() - 1) * num);
        }
    }
    
    public int getProduct(int k) {
        if (pre.size() <= k) return 0;
        return pre.get(pre.size() - 1) / pre.get(pre.size() - k - 1);
    }
    
    private List<Integer> pre = new ArrayList();
}